lib/src/web_socket/w_socket.dart

```diff
@@ -0,0 +1,205 @@
+///
+///     // Receive data
+///     webSocket.listen((data) {
+///       print(data);
+///     });
+///
+/// Finally, you can determine when the web socket connection has been closed
+/// in a few different ways:
+///
+///     Uri uri = Uri.parse('ws://echo.websocket.org');
+///     WSocket webSocket = await WSocket.connect(uri);
+///
+///     // "done" is a Future
+///     webSocket.done.then((_) { ... });
+///
+///     // calling "close()" returns the same Future
+///     webSocket.close().then((_) { ... });
+///
+///     // registering an "onDone()" handler also works
+///     webSocket.listen((_) {}, onDone: () { ... });
+///
+class WSocket extends Stream implements StreamSink {
+
+  /// Create a new WebSocket connection. The given [uri] must use the scheme
+  /// `ws` or `wss`.
+  ///
+  /// Specify the subprotocols the client is willing to speak via [protocols].
+  ///
+  /// Additional headers to be used in setting up the connection can be
+  /// specified in [headers]. This only applies to server-side usage. See
+  /// `dart:io`'s [WebSocket] for more information.
+  static Future<WSocket> connect(Uri uri,
+      {Iterable<String> protocols, Map<String, dynamic> headers}) async {
+    verifyWHttpConfigurationIsSet();
+    var socketController =
+        await common.connect(uri, protocols: protocols, headers: headers);
+    return new WSocket._(socketController);
+  }
+
+  /// Close code set when the WebSocket closes.
+  int _closeCode;
+
+  /// Close reason set when the WebSocket closes.
+  String _closeReason;
+
+  /// Whether or not the web socket is in the process of closing (or already has
+  /// closed). This prevents duplicate behavior if [close] is called multiple times.
+  bool _closing = false;
+
+  /// Completes when..
+  /// - Outgoing communication is closed (sink)
+  /// - Underlying WebSocket connection is closed
+  /// - Incoming communication is closed (stream)
+  Completer _done = new Completer();
+
+  /// Underlying sink used to send WebSocket messages.
+  StreamSink _sink;
+
+  /// Underlying WebSocket instance.
+  dynamic _socket;
+
+  /// Underlying stream of incoming WebSocket messages.
+  Stream _stream;
+
+  /// Private constructor for creating new [WSocket] instances after connecting.
+  WSocket._(WSocketController wSocketController)
+      : _sink = wSocketController.sink,
+        _socket = wSocketController.socket,
+        _stream = wSocketController.stream {
+    wSocketController.done.then((closeEvent) {
+      _closeCode = closeEvent.code;
+      _closeReason = closeEvent.reason;
+      _done.complete();
+    });
+  }
+
+  /// The close code set when the WebSocket connection is closed. If there is
+  /// no close code available this property will be `null`.
+  int get closeCode => _closeCode;
+
+  /// The close reason set when the WebSocket connection is closed. If there is
+  /// no close reason available this property will be `null`.
+  String get closeReason => _closeReason;
```

Comment:
code is definitely null by default, but reason may be an empty string.. I'll check on that.